### PR TITLE
Chameleon PDA's

### PR DESCRIPTION
**Pull request by Rainfey**
## About the PR
Added a chameleon PDA to the chameleon kit
Make chameleon clothing use Slot and EntityWhitelist to select valid variants, rather than Slot and  the WhitelistChameleon tag

## Why / Balance
Major equippable item that was missing from the chameleon kit
Now allows for chameleon clothing items to support subsets of items that fit in their slot (ie you could have Chameleon Magboots which only support magboot variants)

## Technical details
Required adding a handler for PDA's snowflake method of storing their appearance (field on the PDAComponent)
Moved some functionality into SharedChameleonClothing so the server can check if a target is valid
ChameleonClothingComponent and ChameleonBoundUserInterfaceState now contain an EntityWhitelist
Added support for the recent IdentityBlocker partial coverage PR #24741 

## Media
Screenshot 2024-01-30 165959

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: Chameleon PDA

### CODE REVIEW

**Comment by potato1234x** [comment]
the agent id chameleon should probably be fixed instead of removed, knowing the maintainers they probably wont merge until fixed

**Comment by Rainfey** [comment]
Well i wanted to avoid it - but i would also prefer having both chameleon ID and PDA.
Drafting until i rewrite things to support that.
Probably going to add an EntityWhitelist to each chameleon item.

File: Content.Client/Clothing/Systems/ChameleonClothingSystem.cs [hunk 1]
```diff
@@ -47,54 +24,28 @@ protected override void UpdateSprite(EntityUid uid, EntityPrototype proto)
...
-    }
+        if (!component.Default.HasValue)
+            return;
 
-    private void PrepareAllVariants()
-    {
-        _data.Clear();
-        var prototypes = _proto.EnumeratePrototypes<EntityPrototype>();
+        if (prototype == null)
+            prototype = Proto.Index(component.Default.Value);
```
Comment (metalgearsloth) [inline Content.Client/Clothing/Systems/ChameleonClothingSystem.cs:39]: TryIndex

File: Content.Shared/Clothing/Components/ChameleonClothingComponent.cs [hunk 2]
```diff
@@ -38,21 +43,23 @@ public sealed partial class ChameleonClothingComponent : Component
 public sealed class ChameleonBoundUserInterfaceState : BoundUserInterfaceState
 {
     public readonly SlotFlags Slot;
-    public readonly string? SelectedId;
+    public readonly EntProtoId? SelectedId;
+    public readonly EntityWhitelist? Whitelist;
 
-    public ChameleonBoundUserInterfaceState(SlotFlags slot, string? selectedId)
+    public ChameleonBoundUserInterfaceState(SlotFlags slot, EntProtoId? selectedId, EntityWhitelist? whitelist)
```
Comment (metalgearsloth) [inline Content.Shared/Clothing/Components/ChameleonClothingComponent.cs:49]: You can just network it in the component and not need it in the BUI state at all.

File: Content.Shared/Clothing/EntitySystems/SharedChameleonClothingSystem.cs [hunk 3]
```diff
@@ -33,16 +46,31 @@ private void OnGotUnequipped(EntityUid uid, ChameleonClothingComponent component
...
-    protected void UpdateVisuals(EntityUid uid, ChameleonClothingComponent component)
+    protected virtual void UpdateVisuals(EntityUid uid, ChameleonClothingComponent component, EntityPrototype? proto = null)
     {
-        if (string.IsNullOrEmpty(component.Default) ||
-            !_proto.TryIndex(component.Default, out EntityPrototype? proto))
+        if (!component.Default.HasValue)
             return;
 
+        if (proto == null) proto = Proto.Index(component.Default.Value);
+
```
Comment (metalgearsloth) [inline Content.Shared/Clothing/EntitySystems/SharedChameleonClothingSystem.cs:73]: TryIndex

File: Content.Shared/PDA/SharedPdaSystem.cs [hunk 4]
```diff
@@ -62,6 +62,11 @@ private void OnGetAdditionalAccess(EntityUid uid, PdaComponent component, ref Ge
                 args.Entities.Add(id);
         }
 
+        public virtual void UpdatePdaState(EntityUid uid, PdaComponent pda, string newState)
+        {
+            pda.State = newState;
+        }
+
```
Comment (metalgearsloth) [inline Content.Shared/PDA/SharedPdaSystem.cs:69]: This looks unused and it's also undocumented.

File: Content.Shared/Clothing/EntitySystems/SharedChameleonClothingSystem.cs [hunk 5]
```diff
@@ -56,40 +84,130 @@ protected void UpdateVisuals(EntityUid uid, ChameleonClothingComponent component
...
+    /// Get all valid prototypes for a specific chameleon item
+    /// </summary>
+    /// <param name="slot">Which slot does the chameleon item fit in</param>
+    /// <param name="factory">Component factory</param>
+    /// <param name="whitelist">Optional whitelist to match against each prototype</param>
+    /// <returns></returns>
+    public IEnumerable<EntProtoId> GetValidVariants(SlotFlags slot, IComponentFactory factory, EntityWhitelist? whitelist = null)
+    {
+        var outList = new List<EntProtoId>();
+
```
Comment (metalgearsloth) [inline Content.Shared/Clothing/EntitySystems/SharedChameleonClothingSystem.cs:190]: Why not valuelist if you're just going to return a slow ienumerable anyway?

File: Content.Shared/Whitelist/EntityWhitelist.cs [hunk 6]
```diff
@@ -96,5 +97,38 @@ public bool IsValid(EntityUid uid, IEntityManager? entityManager = null)
...
             return false;
         }
+
+        /// <summary>
+        /// Returns whether a given entity prototype fits the whitelist
+        /// </summary>
+        /// <param name="proto">The entity prototype to match</param>
+        /// <returns>True if the prototype matches the whitelist</returns>
+        public bool IsValid(EntityPrototype proto, IComponentFactory factory, IEntityManager? entityManager = null)
+        {
```
Comment (metalgearsloth) [inline Content.Shared/Whitelist/EntityWhitelist.cs:107]: You never use the factory.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Pull request closed without merging**